rosidl_generator_py/resource/_msg_support.c.em

```diff
@@ -248,156 +248,187 @@ nested_type = '__'.join(type_.namespaced_name())
-    Py_ssize_t size = PySequence_Size(field);
-    if (-1 == size) {
-      Py_DECREF(seq_field);
-      Py_DECREF(field);
-      return false;
-    }
-@[      if isinstance(member.type.value_type, AbstractString)]@
-    if (!rosidl_runtime_c__String__Sequence__init(&(ros_message->@(member.name)), size)) {
-      PyErr_SetString(PyExc_RuntimeError, "unable to create String__Sequence ros_message");
-      Py_DECREF(seq_field);
-      Py_DECREF(field);
-      return false;
-    }
-@[      elif isinstance(member.type.value_type, AbstractWString)]@
-    if (!rosidl_runtime_c__U16String__Sequence__init(&(ros_message->@(member.name)), size)) {
-      PyErr_SetString(PyExc_RuntimeError, "unable to create U16String__Sequence ros_message");
-      Py_DECREF(seq_field);
-      Py_DECREF(field);
-      return false;
-    }
-@[      else]@
-    if (!rosidl_runtime_c__@(member.type.value_type.typename)__Sequence__init(&(ros_message->@(member.name)), size)) {
-      PyErr_SetString(PyExc_RuntimeError, "unable to create @(member.type.value_type.typename)__Sequence ros_message");
-      Py_DECREF(seq_field);
-      Py_DECREF(field);
-      return false;
-    }
-@[      end if]@
-    @primitive_msg_type_to_c(member.type.value_type) * dest = ros_message->@(member.name).data;
-@[    else]@
-    Py_ssize_t size = @(member.type.size);
-    @primitive_msg_type_to_c(member.type.value_type) * dest = ros_message->@(member.name);
-@[    end if]@
-    for (Py_ssize_t i = 0; i < size; ++i) {
-@[    if not isinstance(member.type, Array) or not isinstance(member.type.value_type, BasicType) or member.type.value_type.typename not in SPECIAL_NESTED_BASIC_TYPES]@
-      PyObject * item = PySequence_Fast_GET_ITEM(seq_field, i);
-      if (!item) {
-        Py_DECREF(seq_field);
+@[    if isinstance(member.type, AbstractSequence) and isinstance(member.type.value_type, BasicType)]@
+    if (PyObject_CheckBuffer(field)) {
+      // Optimization for converting arrays of primitives
+      Py_buffer view;
+      int rc = PyObject_GetBuffer(field, &view, PyBUF_SIMPLE);
+      if (rc) {
+        PyErr_SetString(PyExc_RuntimeError, "unable to get buffer");
         Py_DECREF(field);
         return false;
       }
+      Py_ssize_t size = view.len / sizeof(@primitive_msg_type_to_c(member.type.value_type));
+      if (!rosidl_runtime_c__@(member.type.value_type.typename)__Sequence__init(&(ros_message->@(member.name)), size)) {
+        PyErr_SetString(PyExc_RuntimeError, "unable to create @(member.type.value_type.typename)__Sequence ros_message");
+        PyBuffer_Release(&view);
+        Py_DECREF(field);
+        return false;
+      }
+      @primitive_msg_type_to_c(member.type.value_type) * dest = ros_message->@(member.name).data;
+      rc = PyBuffer_ToContiguous(dest, &view, view.len, 'C');
+      if (rc) {
+        PyErr_SetString(PyExc_RuntimeError, "unable to copy buffer");
+        PyBuffer_Release(&view);
+        Py_DECREF(field);
+        return false;
```

Comment:
Same question here about whether we should fall back to the slow method.

Reply:
Quote from reference (https://docs.python.org/3/c-api/buffer.html#c.PyBuffer_ToContiguous): "This function fails if len != src->len.".
As I undrestand it - as we just literally took the value from the src it should never happen. However, just for any future possible change in the bahaviour I would left this check, just in case.